Neos.ContentRepository.Core/Classes/Service/ContentStreamPruner.php

```diff
@@ -132,10 +173,159 @@ private function findUnusedAndRemovedContentStreams(): ContentStreams
+            VirtualStreamName::forCategory(ContentStreamEventStreamName::EVENT_STREAM_NAME_PREFIX),
+            EventStreamFilter::create(
+                EventTypes::create(
+                    EventType::fromString('ContentStreamWasCreated'),
+                    EventType::fromString('ContentStreamWasForked'),
+                    EventType::fromString('ContentStreamWasRemoved'),
+                )
+            )
+        );
+
+        /** @var array<string,ContentStreamForPruning> $status */
+        $status = [];
+        foreach ($events as $eventEnvelope) {
+            $domainEvent = $this->eventNormalizer->denormalize($eventEnvelope->event);
+
+            switch ($domainEvent::class) {
+                case ContentStreamWasCreated::class:
+                    $status[$domainEvent->contentStreamId->value] = ContentStreamForPruning::create(
+                        $domainEvent->contentStreamId,
+                        ContentStreamStatus::CREATED,
+                        null
+                    );
+                    break;
+                case ContentStreamWasForked::class:
+                    $status[$domainEvent->newContentStreamId->value] = ContentStreamForPruning::create(
+                        $domainEvent->newContentStreamId,
+                        ContentStreamStatus::FORKED,
+                        $domainEvent->sourceContentStreamId
+                    );
+                    break;
+                case ContentStreamWasRemoved::class:
+                    if (isset($status[$domainEvent->contentStreamId->value])) {
+                        $status[$domainEvent->contentStreamId->value] = $status[$domainEvent->contentStreamId->value]
+                            ->withRemoved();
+                    }
+                    break;
+                default:
+                    throw new \RuntimeException(sprintf('Unhandled event %s', $eventEnvelope->event->type->value));
+            }
+        }
+
+        $workspaceEvents = $this->eventStore->load(
+            VirtualStreamName::forCategory(WorkspaceEventStreamName::EVENT_STREAM_NAME_PREFIX),
+            EventStreamFilter::create(
+                EventTypes::create(
+                    EventType::fromString('WorkspaceWasCreated'),
+                    EventType::fromString('WorkspaceWasDiscarded'),
+                    EventType::fromString('WorkspaceWasPartiallyDiscarded'),
+                    EventType::fromString('WorkspaceWasPartiallyPublished'),
+                    EventType::fromString('WorkspaceWasPublished'),
+                    EventType::fromString('WorkspaceWasRebased'),
+                    // we don't need to track WorkspaceWasRemoved as a ContentStreamWasRemoved event would be emitted before
+                )
+            )
+        );
+        foreach ($workspaceEvents as $eventEnvelope) {
+            $domainEvent = $this->eventNormalizer->denormalize($eventEnvelope->event);
+
+            switch ($domainEvent::class) {
```

Comment:
upsi @kitsunet the `RootWorkspaceWasCreated` event was kind-of ignored here so the content stream was pruned in bernhards case is guess :O ... maybe we should backup the events before or at least ask for confirmation?